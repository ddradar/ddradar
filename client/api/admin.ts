import type { NuxtHTTPInstance } from '@nuxt/http'

import type { SongInfo } from '~/api/song'
import type { NotificationSchema } from '~/core/db/notification'

type NotificationRequest = Omit<
  NotificationSchema,
  'id' | 'sender' | 'timeStamp'
> & {
  id?: string
  timeStamp?: number
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postNotification
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: NotificationRequest
) {
  return $http.$post<NotificationRequest>('/api/v1/admin/notification', {
    ...body,
    sender: 'SYSTEM',
  })
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postSongInfo
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: SongInfo
) {
  return $http.$post<SongInfo>('/api/v1/admin/songs', body)
}
