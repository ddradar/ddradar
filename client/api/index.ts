export const apiPrefix = '/api/v1'
