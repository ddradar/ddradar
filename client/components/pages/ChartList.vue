<template>
  <b-table
    :data="charts"
    striped
    :loading="loading"
    :mobile-cards="false"
    paginated
    per-page="50"
  >
    <template slot-scope="props">
      <b-table-column field="series" label="Series">
        {{ shortenSeries(props.row.series) }}
      </b-table-column>
      <b-table-column field="name" label="Name">
        <nuxt-link
          :to="`/songs/${props.row.id}/${props.row.playStyle}${props.row.difficulty}`"
        >
          {{ props.row.name }}
        </nuxt-link>
      </b-table-column>
      <b-table-column field="difficulty" label="Difficulty">
        <b-tag :type="getTagClass(props.row.difficulty)">
          {{ getDifficulty(props.row.difficulty) }}
        </b-tag>
      </b-table-column>
      <b-table-column field="level" label="Lv" numeric>
        {{ props.row.level }}
      </b-table-column>
      <b-table-column field="level" label="Edit">
        <a
          @click="
            scoreEditorModal(
              props.row.id,
              props.row.playStyle,
              props.row.difficulty
            )
          "
        >
          <b-icon icon="pencil-box-outline" />
        </a>
      </b-table-column>
    </template>

    <template slot="empty">
      <section class="section">
        <div class="content has-text-grey has-text-centered">
          <p>Nothing here.</p>
        </div>
      </section>
    </template>
  </b-table>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import ScoreEditor from '~/components/pages/ScoreEditor.vue'
import {
  ChartInfo,
  getDifficultyName,
  shortenSeriesName,
  SongInfo,
} from '~/types/api/song'

@Component
export default class ChartListComponent extends Vue {
  @Prop({ required: false, type: Array, default: () => [] })
  charts!: ChartInfo[]

  @Prop({ required: false, type: Boolean, default: false })
  loading!: boolean

  shortenSeries(series: string) {
    return shortenSeriesName(series)
  }

  getDifficulty(difficulty: number) {
    return getDifficultyName(difficulty)
  }

  getTagClass(difficulty: number) {
    return `is-${getDifficultyName(difficulty).toLowerCase()}`
  }

  async scoreEditorModal(
    songId: string,
    playStyle: number,
    difficulty: number
  ) {
    const songData = await this.$http.$get<SongInfo>(`/api/v1/songs/${songId}`)
    this.$buefy.modal.open({
      parent: this,
      component: ScoreEditor,
      props: { songId, playStyle, difficulty, songData },
      hasModalCard: true,
      trapFocus: true,
    })
  }
}
</script>
