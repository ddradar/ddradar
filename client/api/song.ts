import type { Api } from '@ddradar/core'
import { Song } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function shortenSeriesName(series: string) {
  return series.replace(/^(DDR |DanceDanceRevolution )\(?([^)]+)\)?$/, '$2')
}

export function getChartTitle({
  playStyle,
  difficulty,
  level,
}: Pick<Api.ChartInfo, 'playStyle' | 'difficulty' | 'level'>) {
  const shortPlayStyle = `${Song.playStyleMap.get(playStyle)![0]}P` as
    | 'SP'
    | 'DP'
  const difficultyName = Song.difficultyMap.get(difficulty)!
  return `${shortPlayStyle}-${difficultyName} (${level})`
}

export function getDisplayedBPM({
  minBPM,
  maxBPM,
}: Pick<Api.SongInfo, 'minBPM' | 'maxBPM'>) {
  if (!minBPM || !maxBPM) return '???'
  if (minBPM === maxBPM) return `${minBPM}`
  return `${minBPM}-${maxBPM}`
}

/**
 * Call "Get Song Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__id/
 */
export function getSongInfo($http: Pick<NuxtHTTPInstance, '$get'>, id: string) {
  return $http.$get<Api.SongInfo>(`${apiPrefix}/songs/${id}`)
}

/**
 * Call "Search Song by Name" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__name__id/
 */
export function searchSongByName(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  nameIndex: number
) {
  return $http.$get<Api.SongListData[]>(`${apiPrefix}/songs/name/${nameIndex}`)
}

/**
 * Call "Search Song by Series" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__series__id/
 */
export function searchSongBySeries(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series: number
) {
  return $http.$get<Api.SongListData[]>(`${apiPrefix}/songs/series/${series}`)
}

/**
 * Call "Search Charts" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/charts__style__level/
 */
export function searchCharts(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  playStyle: Song.PlayStyle,
  level: number
) {
  return $http.$get<Api.ChartInfo[]>(
    `${apiPrefix}/charts/${playStyle}/${level}`
  )
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs--post/
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Api.SongInfo
) {
  return $http.$post<Api.SongInfo>(`${apiPrefix}/songs`, body)
}
