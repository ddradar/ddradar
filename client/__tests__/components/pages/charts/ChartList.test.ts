import type { Api } from '@ddradar/core'
import { testSongData } from '@ddradar/core/__tests__/data'
import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
} from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import VueI18n from 'vue-i18n'

import { getSongInfo } from '~/api/song'
import ChartList from '~/components/pages/charts/ChartList.vue'

jest.mock('~/api/song', () => ({
  ...(jest.requireActual('~/api/song') as object),
  getSongInfo: jest.fn(),
}))
const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(VueI18n)

describe('/components/pages/charts/ChartList.vue', () => {
  describe.each(['ja', 'en'])('{ locale: %s } snapshot test', locale => {
    const i18n = new VueI18n({ locale, silentFallbackWarn: true })
    const stubs = { NuxtLink: RouterLinkStub }
    const wrapper = mount(ChartList, { localVue, stubs, i18n })
    test('{ loading: true } renders loading spin', async () => {
      // Arrange
      wrapper.setProps({ loading: true })
      await wrapper.vm.$nextTick()

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
    test('{ loading: false, charts: [...] } renders chart list', async () => {
      // Arrange
      const charts: Api.ChartInfo[] = [
        {
          id: '9i0q91lPPiO61b9P891O1i86iOP1I08O',
          name: 'EGOISM 440',
          series: 'DanceDanceRevolution (2014)',
          playStyle: 1,
          difficulty: 4,
          level: 19,
        },
        {
          id: 'PO9Pl1q896bDDl89qQb98D80DQoPio1I',
          name: 'ENDYMION',
          series: 'DanceDanceRevolution A',
          playStyle: 1,
          difficulty: 4,
          level: 19,
        },
        {
          id: '1Dl19idl0i0qiqidbDIIbQddiP6o11PP',
          name: 'MAX 360',
          series: 'DanceDanceRevolution A',
          playStyle: 1,
          difficulty: 4,
          level: 19,
        },
        {
          id: '6bid6d9qPQ80DOqiidQQ891o6Od8801l',
          name: 'Over The “Period”',
          series: 'DanceDanceRevolution (2014)',
          playStyle: 1,
          difficulty: 4,
          level: 19,
        },
        {
          id: '606b9d6OiliId69bO9Odi6qq8o8Qd0dq',
          name: 'PARANOiA Revolution',
          series: 'DDR X3 VS 2ndMIX',
          playStyle: 1,
          difficulty: 4,
          level: 19,
        },
        {
          id: '186dd6DQq891Ib9Ilq8Qbo8lIqb0Qoll',
          name: 'Valkyrie dimension',
          series: 'DDR X2',
          playStyle: 1,
          difficulty: 4,
          level: 19,
        },
      ]
      wrapper.setProps({ loading: false, charts })
      await wrapper.vm.$nextTick()

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
    test('{ loading: false, charts: [] } renders empty', async () => {
      // Arrange
      wrapper.setProps({ loading: false, charts: [] })
      await wrapper.vm.$nextTick()

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
  })

  // Method
  describe('scoreEditorModal()', () => {
    const mocks = { $buefy: { modal: { open: jest.fn() } }, $http: {} }
    test('calls $buefy.modal.open()', async () => {
      // Arrange
      mocked(getSongInfo).mockResolvedValue(testSongData)
      const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
      const wrapper = shallowMount(ChartList, { localVue, mocks, i18n })

      // Act
      // @ts-ignore
      await wrapper.vm.scoreEditorModal(
        testSongData.id,
        testSongData.charts[0].playStyle,
        testSongData.charts[0].difficulty
      )

      // Assert
      expect(mocked(getSongInfo)).toBeCalledWith(mocks.$http, testSongData.id)
      expect(mocks.$buefy.modal.open).toBeCalled()
    })
  })
})
