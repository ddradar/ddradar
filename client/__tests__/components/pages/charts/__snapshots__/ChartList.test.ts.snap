// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/pages/charts/ChartList.vue { locale: en } snapshot test { loading: false, charts: [...] } renders chart list 1`] = `
<div class="b-table">
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <div class="table-wrapper">
    <table class="table is-striped">
      <thead>
        <tr>
          <!---->
          <!---->
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Series <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Name <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Difficulty <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap is-numeric"><span class="is-relative"> Lv <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Edit Score <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Series" class="">
            2014
          </td>
          <td data-label="Name" class=""><a>EGOISM 440</a></td>
          <td data-label="Difficulty" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="Lv" class="has-text-right">
            19
          </td>
          <td data-label="Edit Score" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Series" class="">
            A
          </td>
          <td data-label="Name" class=""><a>ENDYMION</a></td>
          <td data-label="Difficulty" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="Lv" class="has-text-right">
            19
          </td>
          <td data-label="Edit Score" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Series" class="">
            A
          </td>
          <td data-label="Name" class=""><a>MAX 360</a></td>
          <td data-label="Difficulty" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="Lv" class="has-text-right">
            19
          </td>
          <td data-label="Edit Score" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Series" class="">
            2014
          </td>
          <td data-label="Name" class=""><a>Over The “Period”</a></td>
          <td data-label="Difficulty" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="Lv" class="has-text-right">
            19
          </td>
          <td data-label="Edit Score" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Series" class="">
            X3 VS 2ndMIX
          </td>
          <td data-label="Name" class=""><a>PARANOiA Revolution</a></td>
          <td data-label="Difficulty" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="Lv" class="has-text-right">
            19
          </td>
          <td data-label="Edit Score" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Series" class="">
            X2
          </td>
          <td data-label="Name" class=""><a>Valkyrie dimension</a></td>
          <td data-label="Difficulty" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="Lv" class="has-text-right">
            19
          </td>
          <td data-label="Edit Score" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <!---->
      </tbody>
      <!---->
    </table>
    <!---->
  </div>
  <div class="top level">
    <div class="level-left"></div>
    <div class="level-right">
      <div class="level-item">
        <nav class="pagination"><a role="button" href="#" disabled="disabled" class="pagination-link pagination-previous"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-left mdi-24px"></i></span></a><a role="button" href="#" disabled="disabled" class="pagination-link pagination-next"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a>
          <ul class="pagination-list">
            <!---->
            <!---->
            <li><a role="button" href="#" aria-current="true" class="pagination-link is-current ">1</a></li>
            <!---->
            <!---->
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`/components/pages/charts/ChartList.vue { locale: en } snapshot test { loading: false, charts: [] } renders empty 1`] = `
<div class="b-table">
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <div class="table-wrapper">
    <table class="table is-striped">
      <thead>
        <tr>
          <!---->
          <!---->
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Series <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Name <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Difficulty <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap is-numeric"><span class="is-relative"> Lv <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Edit Score <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr class="is-empty">
          <td colspan="5">
            <section class="section">
              <div class="content has-text-grey has-text-centered">
                <p>No data</p>
              </div>
            </section>
          </td>
        </tr>
      </tbody>
      <!---->
    </table>
    <!---->
  </div>
  <div class="top level">
    <div class="level-left"></div>
    <div class="level-right">
      <div class="level-item">
        <nav class="pagination"><a role="button" href="#" disabled="disabled" class="pagination-link pagination-previous"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-left mdi-24px"></i></span></a><a role="button" href="#" disabled="disabled" class="pagination-link pagination-next"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a>
          <ul class="pagination-list">
            <!---->
            <!---->
            <!---->
            <!---->
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`/components/pages/charts/ChartList.vue { locale: en } snapshot test { loading: true } renders loading spin 1`] = `
<div class="b-table">
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <div class="table-wrapper">
    <table class="table is-striped">
      <thead>
        <tr>
          <!---->
          <!---->
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Series <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Name <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Difficulty <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap is-numeric"><span class="is-relative"> Lv <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> Edit Score <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr class="is-empty">
          <td colspan="5">
            <section class="section">
              <div class="b-skeleton is-animated">
                <div class="b-skeleton-item is-rounded"></div>
              </div>
              <div class="b-skeleton is-animated">
                <div class="b-skeleton-item is-rounded"></div>
              </div>
              <div class="b-skeleton is-animated">
                <div class="b-skeleton-item is-rounded"></div>
              </div>
            </section>
          </td>
        </tr>
      </tbody>
      <!---->
    </table>
    <transition-stub name="fade">
      <div class="loading-overlay is-active">
        <div class="loading-background"></div>
        <div class="loading-icon"></div>
      </div>
    </transition-stub>
  </div>
  <div class="top level">
    <div class="level-left"></div>
    <div class="level-right">
      <div class="level-item">
        <nav class="pagination"><a role="button" href="#" disabled="disabled" class="pagination-link pagination-previous"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-left mdi-24px"></i></span></a><a role="button" href="#" disabled="disabled" class="pagination-link pagination-next"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a>
          <ul class="pagination-list">
            <!---->
            <!---->
            <!---->
            <!---->
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`/components/pages/charts/ChartList.vue { locale: ja } snapshot test { loading: false, charts: [...] } renders chart list 1`] = `
<div class="b-table">
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <div class="table-wrapper">
    <table class="table is-striped">
      <thead>
        <tr>
          <!---->
          <!---->
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> バージョン <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> 曲名 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> 難易度 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap is-numeric"><span class="is-relative"> レベル <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> スコア編集 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="バージョン" class="">
            2014
          </td>
          <td data-label="曲名" class=""><a>EGOISM 440</a></td>
          <td data-label="難易度" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="レベル" class="has-text-right">
            19
          </td>
          <td data-label="スコア編集" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="バージョン" class="">
            A
          </td>
          <td data-label="曲名" class=""><a>ENDYMION</a></td>
          <td data-label="難易度" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="レベル" class="has-text-right">
            19
          </td>
          <td data-label="スコア編集" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="バージョン" class="">
            A
          </td>
          <td data-label="曲名" class=""><a>MAX 360</a></td>
          <td data-label="難易度" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="レベル" class="has-text-right">
            19
          </td>
          <td data-label="スコア編集" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="バージョン" class="">
            2014
          </td>
          <td data-label="曲名" class=""><a>Over The “Period”</a></td>
          <td data-label="難易度" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="レベル" class="has-text-right">
            19
          </td>
          <td data-label="スコア編集" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="バージョン" class="">
            X3 VS 2ndMIX
          </td>
          <td data-label="曲名" class=""><a>PARANOiA Revolution</a></td>
          <td data-label="難易度" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="レベル" class="has-text-right">
            19
          </td>
          <td data-label="スコア編集" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="バージョン" class="">
            X2
          </td>
          <td data-label="曲名" class=""><a>Valkyrie dimension</a></td>
          <td data-label="難易度" class=""><span class="tag is-challenge"><span class="">CHALLENGE</span>
            <!----></span>
          </td>
          <td data-label="レベル" class="has-text-right">
            19
          </td>
          <td data-label="スコア編集" class=""><a><span class="icon"><i class="mdi mdi-pencil-box-outline mdi-24px"></i></span></a></td>
          <!---->
        </tr>
        <transition-stub name="">
          <!---->
        </transition-stub>
        <!---->
        <!---->
      </tbody>
      <!---->
    </table>
    <!---->
  </div>
  <div class="top level">
    <div class="level-left"></div>
    <div class="level-right">
      <div class="level-item">
        <nav class="pagination"><a role="button" href="#" disabled="disabled" class="pagination-link pagination-previous"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-left mdi-24px"></i></span></a><a role="button" href="#" disabled="disabled" class="pagination-link pagination-next"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a>
          <ul class="pagination-list">
            <!---->
            <!---->
            <li><a role="button" href="#" aria-current="true" class="pagination-link is-current ">1</a></li>
            <!---->
            <!---->
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`/components/pages/charts/ChartList.vue { locale: ja } snapshot test { loading: false, charts: [] } renders empty 1`] = `
<div class="b-table">
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <div class="table-wrapper">
    <table class="table is-striped">
      <thead>
        <tr>
          <!---->
          <!---->
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> バージョン <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> 曲名 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> 難易度 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap is-numeric"><span class="is-relative"> レベル <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> スコア編集 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr class="is-empty">
          <td colspan="5">
            <section class="section">
              <div class="content has-text-grey has-text-centered">
                <p>データがありません</p>
              </div>
            </section>
          </td>
        </tr>
      </tbody>
      <!---->
    </table>
    <!---->
  </div>
  <div class="top level">
    <div class="level-left"></div>
    <div class="level-right">
      <div class="level-item">
        <nav class="pagination"><a role="button" href="#" disabled="disabled" class="pagination-link pagination-previous"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-left mdi-24px"></i></span></a><a role="button" href="#" disabled="disabled" class="pagination-link pagination-next"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a>
          <ul class="pagination-list">
            <!---->
            <!---->
            <!---->
            <!---->
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;

exports[`/components/pages/charts/ChartList.vue { locale: ja } snapshot test { loading: true } renders loading spin 1`] = `
<div class="b-table">
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <!---->
  <div class="table-wrapper">
    <table class="table is-striped">
      <thead>
        <tr>
          <!---->
          <!---->
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> バージョン <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> 曲名 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> 難易度 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap is-numeric"><span class="is-relative"> レベル <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <th draggable="false" class="">
            <div class="th-wrap"><span class="is-relative"> スコア編集 <span class="icon sort-icon is-small is-invisible"><i class="mdi mdi-arrow-up"></i></span></span></div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr class="is-empty">
          <td colspan="5">
            <section class="section">
              <div class="b-skeleton is-animated">
                <div class="b-skeleton-item is-rounded"></div>
              </div>
              <div class="b-skeleton is-animated">
                <div class="b-skeleton-item is-rounded"></div>
              </div>
              <div class="b-skeleton is-animated">
                <div class="b-skeleton-item is-rounded"></div>
              </div>
            </section>
          </td>
        </tr>
      </tbody>
      <!---->
    </table>
    <transition-stub name="fade">
      <div class="loading-overlay is-active">
        <div class="loading-background"></div>
        <div class="loading-icon"></div>
      </div>
    </transition-stub>
  </div>
  <div class="top level">
    <div class="level-left"></div>
    <div class="level-right">
      <div class="level-item">
        <nav class="pagination"><a role="button" href="#" disabled="disabled" class="pagination-link pagination-previous"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-left mdi-24px"></i></span></a><a role="button" href="#" disabled="disabled" class="pagination-link pagination-next"><span class="icon" aria-hidden="true"><i class="mdi mdi-chevron-right mdi-24px"></i></span></a>
          <ul class="pagination-list">
            <!---->
            <!---->
            <!---->
            <!---->
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
